Source/SwiftLintFramework/Rules/QuickDiscouragedCallRule.swift

```diff
@@ -0,0 +1,107 @@
+//
+//  QuickDiscouragedCallRule.swift
+//  SwiftLint
+//
+//  Created by Ornithologist Coder on 8/11/17.
+//  Copyright © 2017 Realm. All rights reserved.
+//
+
+import Foundation
+import SourceKittenFramework
+
+// swiftlint:disable identifier_name
+
+public struct QuickDiscouragedCallRule: ASTRule, OptInRule, ConfigurationProviderRule {
+    public var configuration = SeverityConfiguration(.warning)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "quick_discouraged_call",
+        name: "Quick Discouraged Call",
+        description: "Discouraged call inside 'describe' and/or 'context' block.",
+        kind: .style,
+        nonTriggeringExamples: QuickDiscouragedCallRuleExamples.nonTriggeringExamples,
+        triggeringExamples: QuickDiscouragedCallRuleExamples.triggeringExamples
+    )
+
+    public func validate(file: File,
+                         kind: SwiftExpressionKind,
+                         dictionary: [String: SourceKitRepresentable]) -> [StyleViolation] {
+        guard
+            fileContainsQuickSpec(file: file),
```

Comment:
For the second comment, I'm reviewing the code here and I'm not sure it will generate multiple violations. The code isn't recursively navigating the entire tree. It's only analyzing the calls passed to `describe` && `context`'s trailing closures.

For instance:

```swift
class TotoTests: QuickSpec {
    override func spec() {
        describe("top level describe") {
            let foo = Foo() // violation expected
            let bar = Bar() // violation expected

            context("first level context 1") {
                let foo = Foo() // violation expected

                context("second level context") {
                    beforeEach {
                        let foo = Foo()
                        foo.toto()
                    }

                    it("test") {
                        let foo = Foo()
                        foo.toto()
                    }

                    context("third level context") {
                        let foo = Foo() // violation expected
                    }
                }
            }

            context("first level context 2") {
                let foo = Foo() // violation expected
            }

            context("first level context 3") {
                foo.toto() // violation expected
            }

            context("first level context 4") {
                var foo: Foo!

                beforeEach {
                    foo = Foo()
                }
            }
        }
    }
}
```

When

```swift
public func validate(file: File,
                     kind: SwiftExpressionKind,
                     dictionary: [String: SourceKitRepresentable]) -> [StyleViolation] {
```

is called for the very first time, it's for the top level `describe` which has two parameters. The first one is the block name and second is a closure. Then for this closure we have to navigate thru it looking for calls. The calls are:

* Foo()
* Bar()
* context("first level context 1")
* context("first level context 2")
* context("first level context 3")
* context("first level context 4")

Because `Foo()` and `Bar()` are not `QuickCallKind`, they are added to the violations array.

Then it goes to the next `ASTRule.validate(file:kind:dictionary:)` call for the `context("first level context 1") { }`. Same happens, the name is discarded and the closure is analyzed. For this one, the calls are:

* Foo()
* context("second level context")

and `Foo()` is the only violation in this scope.

For the next `context`, second level context, the calls are:

* beforeEach
* it("test")
* context("third level context") 

and they're all good, so no violations are added to the array.

For the file above, this is the output:

```
$ ./.build/debug/swiftlint lint --path ~/Desktop/quick.swift --enable-all-rules
Loading configuration from '.swiftlint.yml'
Linting Swift files at path /Users/ornithocoder/Desktop/quick.swift
Linting 'quick.swift' (1/1)
/Users/ornithocoder/Desktop/quick.swift:1:1: warning: Explicit Top Level ACL Violation: Top-level declarations should specify Access Control Level keywords explicitly. (explicit_top_level_acl)
/Users/ornithocoder/Desktop/quick.swift:1: warning: File Header Violation: Header comments should be consistent with project patterns. (file_header)
/Users/ornithocoder/Desktop/quick.swift:22:35: warning: Quick Discouraged Call Violation: Discouraged call inside a 'context' block. (quick_discouraged_call)
/Users/ornithocoder/Desktop/quick.swift:8:27: warning: Quick Discouraged Call Violation: Discouraged call inside a 'context' block. (quick_discouraged_call)
/Users/ornithocoder/Desktop/quick.swift:28:27: warning: Quick Discouraged Call Violation: Discouraged call inside a 'context' block. (quick_discouraged_call)
/Users/ornithocoder/Desktop/quick.swift:32:17: warning: Quick Discouraged Call Violation: Discouraged call inside a 'context' block. (quick_discouraged_call)
/Users/ornithocoder/Desktop/quick.swift:4:23: warning: Quick Discouraged Call Violation: Discouraged call inside a 'describe' block. (quick_discouraged_call)
/Users/ornithocoder/Desktop/quick.swift:5:23: warning: Quick Discouraged Call Violation: Discouraged call inside a 'describe' block. (quick_discouraged_call)
Done linting! Found 8 violations, 0 serious in 1 file.
```